### PR TITLE
[chore] refactor mdatagen loader code

### PR DESCRIPTION
**Pull request by codeboten**
This allows us to support all component types via a map. This will be used when adding the Meter func in mdatagen to produce consistent scope names for other components than receivers/scrapers.

### CODE REVIEW

File: cmd/mdatagen/loader.go [hunk 1]
```diff
@@ -267,22 +267,21 @@ func loadMetadata(filePath string) (metadata, error) {
 	return md, nil
 }
 
+var componentTypes = map[string]func(string) string{
+	"connector": func(in string) string { return strings.TrimSuffix(in, "connector") },
+	"exporter":  func(in string) string { return strings.TrimSuffix(in, "exporter") },
+	"extension": func(in string) string { return strings.TrimSuffix(in, "extension") },
+	"processor": func(in string) string { return strings.TrimSuffix(in, "processor") },
+	"scraper":   func(in string) string { return strings.TrimSuffix(in, "scraper") },
+	"receiver":  func(in string) string { return in },
```
Comment (dmitryax) [inline cmd/mdatagen/loader.go:276]: Why don't we trim the suffix for `receiver` anymore?

**Review by dmitryax — COMMENTED** [review]
This not only refactoring, right? we need to introduce scope for other types?

File: cmd/mdatagen/loader.go (hunk 1 above)
Comment (codeboten) [reply cmd/mdatagen/loader.go:276]: the receiver suffix was never trimmed. i added a comment about this in the code below, i recall this was for backwards compatibility

**Comment by codeboten** [comment]
> This not only refactoring, right? we need to introduce scope for other types?

@dmitryax right, this will introduce scope for other components, which is then used in this PR https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29927

File: cmd/mdatagen/loader.go (hunk 1 above)
Comment (dmitryax) [reply cmd/mdatagen/loader.go:276]: But this is used for label creation as well in README.md like `receiver/kubeletstats` not `receiver/kubeletstatsreceiver`

Comment (codeboten) [reply cmd/mdatagen/loader.go:276]: the `func` specified in this map is only used in `scopeName` below, which I thought was only used in the metrics template to set the scope name. Are you saying this scopeName is used in other places as well? Note that this change does not change how receiver is trimmed for `shortFolderName` which is used in other places.

Comment (codeboten) [reply cmd/mdatagen/loader.go:276]: This is why `make generate` doesn't show any diffs

**Review by bogdandrutu — APPROVED** [review]

File: cmd/mdatagen/loader.go (hunk 1 above)
Comment (dmitryax) [reply cmd/mdatagen/loader.go:276]: Right.  I missed that the `shortFolderName` func uses the keys only.

**Pull request merged**